### ISSUE TITLE
Query planner: Support LIMIT of GROUP BY queries without ORDER BY

### ISSUE THREAD

**Issue by henrikingo** [crate/crate#16827]
### Problem Statement

Suppose a new user is playing with CrateDB, and executes a simple query like the following:

    SELECT *
    FROM iot_events
    WHERE date > "2014-10-09" AND date <= 2024-10-09
    GROUP BY date

For a large database this will result in hundreds of thousands of groups and as it doesn't fit in RAM, the group by will fail.

You might then think adding a LIMIT clause will help, but it doesn't:

    ...
    GROUP BY date
    LIMIT 10

The problem is that LIMIT and OFFSET  only makes sense when they can operate on a sorted result set. However, adding ORDER BY to this query won't help reduce the memory consumption of the GROUP BY. 

### Possible Solutions

Hernan pointed out that it would nevertheless be possible to apply the LIMIT to the GROUP BY. The idea would be to just randomly compute the first 10 groups that happen to arrive, and discard everything else.

Note that supporting OFFSET is meaningless. Since the data is not in any particular order, returning the same N records for both of the below queries would be correct:

    ...
    GROUP BY date
    LIMIT 10

    ...
    GROUP BY date
    LIMIT 20 OFFSET 10

It is necessary to coordinate between nodes so that all nodes collect the same set of groups.

### Considered Alternatives

The counter argument to this feature is that it doesn't make sense to support LIMIT in any query where the results ren't sorted. For something like showing the results in pages, it wouldn't be able to do that, due to lack of ORDER BY.

However, I'm filing this ticket for discussion / estimation, because supporting LIMIT does have some value for a new user exploring CrateDB, or even an experienced user who is building a query or exploring an unknown data set.

**Comment by mfussenegger** [comment]
Just to add: we already do something like this if there are no aggregations present:

https://github.com/crate/crate/pull/9371

**Comment by henrikingo** [comment]
Correct. In this case the goal is not to terminate early but to fit in memory and allow the query to complete at all.

**Comment by BaurzhanSakhariev** [comment]
Should be an optimization rule (which can be turned off) or something that we always apply?

**Comment by henrikingo** [comment]
@BaurzhanSakhariev Why would you want to disable this?

**Comment by matriv** [comment]
I guess, to follow the pattern about planner optimizer rules, that each one of them can be disabled if needed, in case there is some edge case which breaks if the rule is applied.
